### ISSUE TITLE
overlap backward and optimization in TensorFlow 2.0

### ISSUE THREAD

**Issue by QiJune** [tensorflow/tensorflow#33274]
**System information**
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

A typical training process in TensorFlow 2.0 could be as following:

```python
def train(model, dataset, optimizer):
  for x, y in dataset:
    with tf.GradientTape() as tape:
      prediction = model(x)
      loss = loss_fn(prediction, y)
    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
```
First, it does backward computation and gets all gradients. Then, the gradients will be updated in local, or send to a remote node to do optimization. The backward computation and gradient optimization are sequential.

In a distributed training scenario, sending gradients will be time-consuming. It's better to overlap communication and computation. Once one layer completes its backward computation, it could start sending its gradients out, and do optimization.

I am wondering if there is any approach to achieve such a goal.

Thanks!

**Will this change the current api? How?**

Or could we expose a callback interface, to let users insert customized processing logic after a gradient is calculated?

There are different cases:

- Case 1, optimization in the remote parameter server

```python
gradients = tape.gradient(loss, model.trainable_variables, send_callback)
```

- Case 2, gradients need to be averaged by allreduce operation across nodes.

```python
gradients = tape.gradient(loss, model.trainable_variables, allreduce_callback)
optimizer.apply_gradients(zip(gradients, model.trainable_variables))
```

- Case 3, gradients are applied in local

```python
gradients = tape.gradient(loss, model.trainable_variables, None)
optimizer.apply_gradients(zip(gradients, model.trainable_variables))
```

**Who will benefit with this feature?**

**Any Other info.**

**Comment by guptapriya** [comment]
Hi - if you use tf.function around your gradient computation and aggregation, then the underlying graph can often do these optimizations for you. 
I believe they do happen in certain circumstances where it makes sense. But sometimes, combining gradients from multiple layers and doing an all-reduce together is more efficient. Have you noticed an actual performance issue from this? Our goal is to make such optimizations transparent to the user. 

cc @yuefengz @dubey

**Comment by QiJune** [comment]
@guidj  Thanks for your reply.

I notice that there is an Async mode in [EagerExecutor](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/eager/eager_executor.h#L95) in the C++ runtime of TF2.0 Eager version. It launches EagerOperation asynchronously. In addition, there is a `RemoteCopyNode` which supports "remote-->remote"/"remote-->local"/"local-->remote" copy. The node could be launched by EagerExecutor. So, in C++ runtime, this problem is well addressed. The computation and communication are overlapped.

However, we want to achieve such effects in Python without touching or modifying C++ runtime. One way is to support inserting an end callback function/or operator if possible to a backward operator. Once the backward operator finishes its backward computation, the callback function will be launched. Thus, if we could insert `send_gradient`  or `allreduce_gradient` end callback to a backward operator, the computation and communication will be overlapped simply with Python API.

Following is the logic which gets the gradient function of an operator:
https://github.com/tensorflow/tensorflow/blob/7b80146babce4a998e4654f5f55e09c3e6f1144d/tensorflow/python/eager/backprop.py#L119-L141

Maybe we could add another callback argument here.

I am not sure if I find the right place to add such logic. I am willing to work on this if the callback mechanism is reasonable.

**Comment by QiJune** [comment]
@guptapriya   Thanks for your reply. 

In some large model training, such as a model with a big embedding table out of a single computer memory, we usually use parameter server distributed strategy. The gradients are pushed to the parameter servers, and then aggregated and applied to parameters stored in parameter servers.

I notice that there is an Async mode in [EagerExecutor](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/eager/eager_executor.h#L95) in the C++ runtime of TF2.0 Eager version. It launches EagerOperation asynchronously. In addition, there is a `RemoteCopyNode` which supports "remote-->remote"/"remote-->local"/"local-->remote" copy. The node could be launched by EagerExecutor. So, in C++ runtime, this problem is well addressed. The computation and communication are overlapped.

However, we want to achieve such effects in Python without touching or modifying C++ runtime.

One way is to support inserting an end callback function/or operator if possible to a backward operator. Once the backward operator finishes its backward computation, the callback function will be launched. Thus, if we could insert `send_gradient`  or `allreduce_gradient` end callback to a backward operator, the computation and communication will be overlapped simply with Python API.

Following is the logic which gets the gradient function of an operator:
https://github.com/tensorflow/tensorflow/blob/7b80146babce4a998e4654f5f55e09c3e6f1144d/tensorflow/python/eager/backprop.py#L119-L141

Maybe we could add another callback argument here.

I am not sure if I find the right place to add such logic. I am willing to work on this if the callback mechanism is reasonable.

**Comment by guptapriya** [comment]
You should be able to use the async mechanism without having to modify the c++ runtime. @qqfish - is async the default execution mode now in 2.0? If not, what's the python API to enable it?

Also, relatedly, @zongweiz is working on exploring ways of overlapping gradient computation and communication for multi worker sync training in TF 2.0.

**Comment by qqfish** [comment]
For distributed training, it is enabled by default in TF 2.1

**Comment by saikumarchalla** [comment]
@QiJune Could you please check the above comments. If it is still an issue,please let us know. Thanks!
